nomad/alloc_endpoint.go

```diff
@@ -445,3 +445,138 @@ func (a *Alloc) GetServiceRegistrations(
 		},
 	})
 }
+
+// GetIdentities allows nodes to retrieve workload identities for their
+// allocations.
+//
+// This is an internal-only RPC and not exposed via the HTTP API.
+func (a *Alloc) GetIdentities(args *structs.AllocIdentitiesRequest, reply *structs.AllocIdentitiesResponse) error {
+
+	authErr := a.srv.Authenticate(a.ctx, args)
+
+	// Ensure the connection was initiated by a client if TLS is used.
+	if err := validateTLSCertificateLevel(a.srv, a.ctx, tlsCertificateLevelClient); err != nil {
+		return err
+	}
+	if done, err := a.srv.forward("Alloc.GetIdentities", args, args, reply); done {
+		return err
+	}
+	a.srv.MeasureRPCRate("alloc", structs.RateMetricRead, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	defer metrics.MeasureSince([]string{"nomad", "alloc", "get_identities"}, time.Now())
+
+	if len(args.Identities) == 0 {
+		// Client bug
+		return fmt.Errorf("no identities requested")
+	}
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, state *state.StateStore) error {
+			// Lookup the allocations
+			thresholdMet := false
+			for _, idReq := range args.Identities {
+				out, err := state.AllocByID(ws, idReq.AllocID)
+				if err != nil {
+					return err
+				}
+
+				if out == nil {
+					// Alloc may have been GC'd and therefore should not be able to get
+					// new identities.
+					continue
+				}
+
+				if out.ModifyIndex > args.QueryOptions.MinQueryIndex {
```

Comment:
Good catch! This is just a straight up bug because I copied the "normal" stale/waitindex/blockingquery code!

I should just be checking that the `allocs` table index is >= min index and remove the -1 from `WaitIndex: alloc.CreateIndex-1` in widmgr. That will make the RPC block until the alloc has been created. If it hasn't at that point then either something fishy is going on, or the alloc has been GC'd and in either case we shouldn't sign its identities.

Reply:
(The `-1` bit _is_ necessary because blocking queries look for the index _after_ what you request, but Modify instead of Create was still a bug.)